Circle CI build job for ARM64

odidev opened the pull request:
**What type of PR is this?**

/kind feature


**Any specific area of the project related to this PR?**

/area build

**What this PR does / why we need it**:
Added Circle CI job to build Falco for ARM64 platform. 

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #1891 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1997#" title="Author self-approved">odidev</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jonahjon** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @jonahjon` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[.circleci/OWNERS](https://github.com/falcosecurity/falco/blob/master/.circleci/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jonahjon"]} -->

poiana commented:
Welcome @odidev! It looks like this is your first PR to falcosecurity/falco 🎉

FedeDP commented on .circleci/config.yml:
```diff
@@ -1,5 +1,29 @@
+    machine:
+      enabled: true
+      image: ubuntu-2004:202101-01
+    resource-class: arm.medium
```
This should be `resource_class`.

FedeDP commented on .circleci/config.yml:
```diff
@@ -357,11 +381,13 @@ jobs:
           command: |
             FALCO_VERSION=$(cat /build/release/userspace/falco/config_falco.h | grep 'FALCO_VERSION ' | cut -d' ' -f3 | sed -e 's/^"//' -e 's/"$//')
             /source/falco/scripts/publish-rpm -f /build/release/falco-${FALCO_VERSION}-x86_64.rpm -r rpm-dev
+            /source/falco/scripts/publish-rpm -f /build/release/falco-${FALCO_VERSION}-aarch64.rpm -r rpm-dev
```
The same lines must be added also to `publish/packages` and `publish/packages-deb` job, again except for the bin static one.

FedeDP commented:
So, we surely need to port  `publish-deb` script.
As you can see, `publish-bin` already takes architecture as argument. I think we must do the same with `publish-deb`:
basically, on this line: https://github.com/falcosecurity/falco/blob/master/scripts/publish-deb#L41 you must check the external argument provided instead of declaring a local variable.

`publish-rpm` does not seem to reference `x86_64` or any `uname` variable, therefore i guess we shouldn't touch it.

I will need someone else to review the workflow too :) It seems fine to me:
## Build and test workflow
* we have the `build-arm64` job that builds and stores artifacts for arm64
* `build-arm64` is required by `rpm-sign` that is itself required by `publish/packages-dev`
* `build-arm64` is also required by `publish/packages-deb-dev`

## Release workflow
* we have the `build-arm64` job that builds and stores artifacts for arm64
* `build-arm64` is required by `rpm-sign` that is itself required by `publish/packages`
* `build-arm64` is also required by `publish/packages-deb`

FedeDP commented:
To maintainers: note that the `build-arm64` job will never work unless the `falcosecturity/falco-builder:latest` image is available for arm64; that's part of #1990!

FedeDP commented:
@odidev it seems like the new v2.1 workflows version is breaking the build:
> # [#/jobs/driver-loader~1integration] string [driver-loader/integration] does not match pattern ^[A-Za-z][A-Za-z\\s\\d_-]*$

Can you revert to 2?

FedeDP commented:
Given that this PR depends upon #1990 , if you wish i can take over the `publish` scripts part; then you will need to rebase this one, once my PR is reviewed and eventually merged.
WDYT?

odidev commented:
> Can you revert to 2? 

Usage of arm resources makes it mandatory to use workflow 2.1 . I checked out the error its nothing major, changing the name of the job should do the trick, I think it would better to mark this as WIP as it still might need some more modifications based on version 2.1 which can be figured only when we are able to run it. We can focus on releasing docker images and binary for arm64 for now. 

> Moreover, can you set "Fixes [#1891](https://github.com/falcosecurity/falco/issues/1891)" in the OP? 

The issue is for getting testing enabled for arm64, currently we are skipping it. Once we are able to add it too we can probably close it.

> Given that this PR depends upon [#1990](https://github.com/falcosecurity/falco/pull/1990) , if you wish i can take over the publish scripts part; then you will need to rebase this one, once my PR is reviewed and eventually merged. 
WDYT? 

Sure thing.

FedeDP commented:
> The issue is for getting testing enabled for arm64

Mmh aren't we actually testing it? https://github.com/falcosecurity/falco/pull/1997/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R14

> Usage of arm resources makes it mandatory to use workflow 2.1

Oh i see thanks! Yeah renaming the job should do the trick; i think you can proceed with that!

FedeDP commented:
@odidev i updated #1990 fixing publishing (deb,rpm) scripts.
They must be used like this:
```
/source/falco/scripts/publish-rpm -f /build/release/falco-${FALCO_VERSION}-x86_64.rpm -f /build/release/falco-${FALCO_VERSION}-aarch64.rpm -r rpm

/source/falco/scripts/publish-deb -f /build/release/falco-${FALCO_VERSION}-x86_64.deb -f /build/release/falco-${FALCO_VERSION}-aarch64.deb -r deb-dev
```

(notice that passing both packages in a single call is needed!)
Can you update your PR accordingly?
Thanks!

odidev commented:
> Mmh aren't we actually testing it? https://github.com/falcosecurity/falco/pull/1997/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R14 

Ah ok updated. 

> Oh i see thanks! Yeah renaming the job should do the trick; i think you can proceed with that! 

I have changed the name of the job and validated configuration file using CircleCI command line tool as well. 
**Note:** I had to change name of all jobs as `/` was present in the name of jobs which generates error in workflow 2.1 as CircleCI that it is not able to find orb of the name same as letters before the `/`. 

> (notice that passing both packages in a single call is needed!) 
Can you update your PR accordingly? 
Thanks! 

I have updated it accordingly.

FedeDP commented:
Wow thank you! This PR should be ready once my #1990 is merged!
Thank you for your really hard work!!

odidev commented:
Since this PR depends on #1990 , by when can we expect it to get merged?

leogr commented:
/milestone 0.33.0

jasondellaluce commented:
#1990 has been merged! 🥳 @odidev, would you proceed and rebase this on master when you have some spare time? Looking forward to see this getting merged too!

leogr commented:
Hey @odidev 

I'm going to rebase your PR, because we need it to fix the CI :) 
I'm also going to rebuild and push the `falcosecurity/falco-builder` image which is need to be updated to build for arm64, as reported by @FedeDP in [this comment](https://github.com/falcosecurity/falco/pull/1997#issuecomment-1122544714).

I'll keep you posted.

leogr commented:
Quick update: we're trying to test the new `falcosecurity/falco-builder` with this PR. So, sorry if we're spamming you :sweat_smile: We hope are close to a fix.

poiana commented:
LGTM label has been added.  <details>Git tree hash: d70248bc5e26c4eb8d596668f04f3be003429cef</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1997#pullrequestreview-1000162915" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/falco/pull/1997#" title="Author self-approved">odidev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[.circleci/OWNERS](https://github.com/falcosecurity/falco/blob/master/.circleci/OWNERS)~~ [jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

leogr commented:
Closing and reopening to trigger the CI
/close

poiana commented:
@leogr: Closed this PR.

<details>

In response to [this](https://github.com/falcosecurity/falco/pull/1997#issuecomment-1150714565):

>Closing and reopening to trigger the CI
>/close


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leogr commented:
/reopen

poiana commented:
@leogr: Reopened this PR.

<details>

In response to [this](https://github.com/falcosecurity/falco/pull/1997#issuecomment-1150714642):

>/reopen


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leogr commented:
Closing and reopening to update CI required checks
/close

poiana commented:
@leogr: Closed this PR.

<details>

In response to [this](https://github.com/falcosecurity/falco/pull/1997#issuecomment-1150726614):

>Closing and reopening to update CI required checks
>/close


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leogr commented:
/reopen

poiana commented:
@leogr: Reopened this PR.

<details>

In response to [this](https://github.com/falcosecurity/falco/pull/1997#issuecomment-1150726702):

>/reopen


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
